flepimop/gempyor_pkg/src/gempyor/_pydantic_ext.py

```diff
@@ -1,22 +1,84 @@
-from typing import Any
+"""
+Gempyor specific Pydantic extensions.
 
-__all__ = ["_ensure_list", "_override_or_val"]
+This module contains functions that are useful for working with and creating Pydantic
+models.
+"""
 
+__all__ = ()
 
-# once 3.12, use type parametrization
-# def _ensure_list[T](value: T | list[T] | tuple) -> list[T]:
-def _ensure_list(value: Any) -> list | None:
+
+from typing import TypeVar, overload
+
+
+T = TypeVar("T")
+U = TypeVar("U")
+
+
+def _ensure_list(value: list[T] | tuple[T] | T | None) -> list[T] | None:
+    """
+    Ensure that a list, tuple, or single value is returned as a list.
+    
+    Args:
+        value: A value to ensure is a list.
+    
+    Returns:
+        A list of the value(s), if the `value` is not None.
+
+    Examples:
+        >>> from gempyor._pydantic_ext import _ensure_list
+        >>> _ensure_list(None) is None
+        True
+        >>> _ensure_list("abc")
+        ['abc']
+        >>> _ensure_list(123)
+        [123]
+        >>> _ensure_list([True, False, True])
+        [True, False, True]
+        >>> _ensure_list((True, False, True))
+        [True, False, True]
+        >>> _ensure_list((1, 2, 3, 4))
+        [1, 2, 3, 4]
+    """
     if value is None:
         return None
-    elif isinstance(value, list):
+    if isinstance(value, list):
         return value
-    elif isinstance(value, tuple):
+    if isinstance(value, tuple):
         return list(value)
-    else:
-        return [value]
+    return [value]
+
+
+@overload
+def _override_or_val(override: T, value: U | None) -> T: ...
```

Comment:
Hmm. So value should be a T and *not* a None, and override should be T | None, and should always return a T.

If we're dealing with extensions, then yes, could have a U subclass of T, and return a U when override is not None

Reply:
I'll adjust docs and tests accordingly.